### PR TITLE
Per machine 'build_' and '' options

### LINKED ISSUE #4963: RFC: Proposal to rework c_args and friends
The behavior of `*_args` is currently inconsistent between native and cross builds. In native builds, it affects both the intermediate "native" tools that we want to run at build time, and the actual end-goal build targets. In a cross build, the later become "cross" targets and thus are not affected by `*_args` (except those added in a cross file).

There are a couple of problems with this:

- It's not obvious what is going on and why. Issues have been opened because of this confusion in the past. See #4767, #4331.
- There is no universal way to configure extra arguments of build targets, short of introducing a custom option.
- There isn't even a way to globally add custom cross arguments (again, without modifying the build definitions or the cross file).
- Using the same flags for intermediate build-tools as for the actual targets may not be what you want, even if not cross-compiling.

What I'm suggesting is the same concept outlined in #4933: think in terms of `build` and `host`, not `native` and `cross`. So instead of `c_args` etc., we could have something like this:

- `build_<lang>_args`, `build_<lang>_link_args`: used for "native" targets only (`native` is explicitly `true`).
- `host_<lang>_args`, `host_<lang>_link_args`: used for other targets (`native` is either `false` or unspecified).
- `<lang>_args` set in the cross files may map to `host_<lang>_args`.
- In the future, maybe also deprecate the `native` kwarg and replace it with `machine`, accepting one of the global machine objects.

### PR DESCRIPTION
**Pull request by Ericson2314**
See feature request #4963 for what this does and why.

I'm open to suggestions on how to make the code better, though I suspect very nice stuff would have to wait on a more general cleanup of options (#4574).

TODO:
 - [x] Make sure passes existing tests
 - [x] Pull test case from https://github.com/mesonbuild/meson/pull/5255 once that is merged to 0.50, and make sure passes

Closes #4963, #5097, #4331 #4331, #1040

CC @dcbaker @nirbheek @Akaricchi

### CODE REVIEW

File: mesonbuild/coredata.py [hunk 1]
```diff
@@ -418,24 +420,50 @@ def set_buildtype_from_others(self):
...
 
-    def _get_all_nonbuiltin_options(self):
-        yield self.backend_options
-        yield self.user_options
-        yield from self.get_all_compiler_options()
-        yield self.base_options
+    @staticmethod
+    def get_prefixed_options_per_machine(
+        options_per_machine # : PerMachine[Dict[str, _V]]]
+    ) -> Dict[str, _V]:
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [F821] undefined name '_V' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-3979932))

<sub>posted by Sider</sub>

File: mesonbuild/coredata.py [hunk 2]
```diff
@@ -418,24 +420,50 @@ def set_buildtype_from_others(self):
...
+            for for_machine, prefix in machines_and_prefixes
+            for k, v in options_per_machine[for_machine].items()
+        }
+
+    @staticmethod
+    def get_unprefixed_options_per_machine_keyset(
+        options_per_machine # : PerMachine[Dict[str, _V]]
+    ) -> FrozenSet[str]:
+        return frozenset(options_per_machine.build) | \
+                frozenset(options_per_machine.host) | \
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [E126] continuation line over-indented for hanging indent ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-3979933))

<sub>posted by Sider</sub>

File: mesonbuild/coredata.py [hunk 3]
```diff
@@ -418,24 +463,50 @@ def set_buildtype_from_others(self):
...
 
-    def _get_all_nonbuiltin_options(self):
-        yield self.backend_options
-        yield self.user_options
-        yield from self.get_all_compiler_options()
-        yield self.base_options
+    @staticmethod
+    def get_prefixed_options_per_machine(
+        options_per_machine # : PerMachine[Dict[str, _V]]]
+    ) -> Dict[str, _V]:
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [F821] undefined name '_V' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-3981229))

<sub>posted by Sider</sub>

File: mesonbuild/coredata.py [hunk 4]
```diff
@@ -418,24 +464,53 @@ def set_buildtype_from_others(self):
...
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [F821] undefined name '_V' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-3981247))

<sub>posted by Sider</sub>

File: mesonbuild/coredata.py [hunk 5]
```diff
@@ -509,13 +587,18 @@ def set_default_options(self, default_options, subproject, env):
         # Some options default to environment variables if they are
         # unset, set those now. These will either be overwritten
         # below, or they won't.
-        options['pkg_config_path'] = os.environ.get('PKG_CONFIG_PATH', '').split(':')
+        if not 'pkg_config_path' in env.cmd_line_options:
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [E713] test for membership should be 'not in' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-3981248))

<sub>posted by Sider</sub>

File: mesonbuild/coredata.py (hunk 4 above)
Comment (Ericson2314) [reply mesonbuild/coredata.py]: I'm not sure what to do about this? `flake8` is just wrong, the class member is defined just above. (and yes I'm developing with `3.5`.)

**Comment by jpakkane** [comment]
I don't like that this causes an option multiplication. We already have too many of them.

**Comment by Ericson2314** [comment]
@jpakkane In what sense? There is the overlap of `(build|host)_foo` and `foo`, but there isn't actually any more *state*: compiler options and the pkg config path were already being handled for platform. I'd personally be fine with getting rid of `foo` and requiring `(build|host)_foo`, but I figured that would be too rough a transition without a deprecation cycle.

**Comment by Ericson2314** [comment]
The MSVC jobs fail because `b_lto` isn't a flag. But how/why did that work before when it's only for GNU compilers?

**Comment by Ericson2314** [comment]
After wresting with Windows VMs to get a working environment I think I've finally fixed all tests. Let's discuss this feature / native-cross vs build-host. Maybe it will take real-time discussion, but I hope we can all come to a common understanding.

**Comment by Ericson2314** [comment]
OK whew had to change priority order on some options, but now it works.

**Comment by Ericson2314** [comment]
Rebased because another PR of mine was merged. Should still passed tests.

File: mesonbuild/coredata.py [hunk 6]
```diff
@@ -680,19 +773,21 @@ def parse_cmd_line_options(args):
...
-        for name in names:
-            value = getattr(args, name, None)
-            if value is not None:
-                if name in args.cmd_line_options:
-                    cmdline_name = BuiltinOption.argparse_name_to_arg(name)
-                    raise MesonException(
-                        'Got argument {0} as both -D{0} and {1}. Pick one.'.format(name, cmdline_name))
-                args.cmd_line_options[name] = value
-                delattr(args, name)
+    for name, builtin in chain(
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [B007] Loop control variable 'builtin' not used within the loop body. If this is intended, start the name with an underscore. ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4029502))

<sub>posted by Sider</sub>

**Comment by Ericson2314** [comment]
Ping.

File: mesonbuild/envconfig.py [hunk 7]
```diff
@@ -16,7 +16,11 @@
 import typing
 
 from . import mesonlib
-from .mesonlib import EnvironmentException, MachineChoice, PerMachine
+from .mesonlib import (
```
Comment (ignatenkobrain) [inline mesonbuild/envconfig.py]: __[Flake8]__ [F401] '.mesonlib.MachineChoice' imported but unused ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4059423))

<sub>posted by Sider</sub>

File: mesonbuild/coredata.py [hunk 8]
```diff
@@ -26,7 +26,7 @@
 import ast
 import argparse
 import configparser
-from typing import Optional, Any, TypeVar, Generic, Type, List
+from typing import Optional, Any, Dict, FrozenSet, TypeVar, Generic, Type, List, Iterable
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [F401] 'typing.FrozenSet' imported but unused ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4059422))

<sub>posted by Sider</sub>

File: mesonbuild/coredata.py [hunk 9]
```diff
@@ -354,28 +360,40 @@ def init_backend_options(self, backend_name):
...
+        opt.set_value(value)
         # Make sure that buildtype matches other settings.
         if optname == 'buildtype':
             self.set_others_from_buildtype(value)
         else:
             self.set_buildtype_from_others()
+        return True
+
+    def set_builtin_option(self, optname, value):
+        res = _try_set_builtin_option(self, optname, value)
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [F821] undefined name '_try_set_builtin_option' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4063459))

<sub>posted by Sider</sub>

File: mesonbuild/mintro.py [hunk 10]
```diff
@@ -230,7 +230,12 @@ def list_buildoptions(coredata: cdata.CoreData) -> List[dict]:
     add_keys(optlist, coredata.backend_options, 'backend')
     add_keys(optlist, coredata.base_options, 'base')
     # TODO others
-    add_keys(optlist, coredata.compiler_options.build, 'compiler')
+    add_keys(optlist,coredata.compiler_options.host,'compiler (for host machine)')
```
Comment (ignatenkobrain) [inline mesonbuild/mintro.py]: __[Flake8]__ [E231] missing whitespace after ',' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4065126))

<sub>posted by Sider</sub>

File: mesonbuild/rewriter.py [hunk 11]
```diff
@@ -467,7 +467,8 @@ def process_default_options(self, cmd):
             **cdata.builtins,
             **cdata.backend_options,
             **cdata.base_options,
-            **cdata.compiler_options.build,
+            **cdata.compiler_options.host,
+            **{'build_' + k: o for k, o in coredata.compiler_options.build.items()}
```
Comment (ignatenkobrain) [inline mesonbuild/rewriter.py]: __[Flake8]__ [F821] undefined name 'coredata' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4065125))

<sub>posted by Sider</sub>

File: mesonbuild/mintro.py [hunk 12]
```diff
@@ -229,8 +229,12 @@ def list_buildoptions(coredata: cdata.CoreData) -> List[dict]:
     add_keys(optlist, core_options, 'core')
     add_keys(optlist, coredata.backend_options, 'backend')
     add_keys(optlist, coredata.base_options, 'base')
-    # TODO others
-    add_keys(optlist, coredata.compiler_options.build, 'compiler')
+    add_keys(optlist,coredata.compiler_options.host,'compiler (for host machine)')
```
Comment (ignatenkobrain) [inline mesonbuild/mintro.py]: __[Flake8]__ [E231] missing whitespace after ',' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4065127))

<sub>posted by Sider</sub>

File: mesonbuild/mintro.py [hunk 13]
```diff
@@ -229,8 +229,12 @@ def list_buildoptions(coredata: cdata.CoreData) -> List[dict]:
     add_keys(optlist, core_options, 'core')
     add_keys(optlist, coredata.backend_options, 'backend')
     add_keys(optlist, coredata.base_options, 'base')
-    # TODO others
-    add_keys(optlist, coredata.compiler_options.build, 'compiler')
+    add_keys(optlist,coredata.compiler_options.host, 'compiler (for host machine)')
```
Comment (ignatenkobrain) [inline mesonbuild/mintro.py]: __[Flake8]__ [E231] missing whitespace after ',' ([link](https://sider.review/gh/repos/19784232/pulls/5263#issue-4065128))

<sub>posted by Sider</sub>

**Comment by Ericson2314** [comment]
Thanks for reviewing those other PRs @jpakkane. Rebased this now that they are merged—this should be good to go and indeed contains the test you want.

File: test cases/unit/55 pkg_config_path option/meson.build [hunk 14]
```diff
@@ -1,3 +1,12 @@
 project('pkg_config_path option')
 
-dependency('totally_made_up_dep', method : 'pkg-config')
+build = dependency('totally_made_up_dep', native: true, method : 'pkg-config')
+host = dependency('totally_made_up_dep', native: false, method : 'pkg-config')
+
+# TODO always test we can do this separately
```
Comment (Ericson2314) [inline test cases/unit/55 pkg_config_path option/meson.build]: #4010 fixes this.

**Comment by jpakkane** [comment]
Before diving deeper I'd like to point out that we have not preserved `build_` prefix for option names in any release (I just merged it to trunk). It's unclear whether we can just grab it like this without giving people even a single release to prepare.

Is there a way to easily grep existing projects to see if people are using option names that start with `build_`? Maybe from @jeandet 's CI setup?

**Comment by Ericson2314** [comment]
That's a good point. It was for reasons I like this that I proposed hierarchical option names in https://github.com/mesonbuild/meson/issues/4574#issuecomment-493716480.

**Comment by Ericson2314** [comment]
Oh it just occurred to me that this doesn't take the `build_` prefix across the board. It only takes `build_pkg_config_path` and `build_<compiler-option>`. That's a few new flags, but a far cry than all `build_*`.

**Comment by jpakkane** [comment]
Having some sort of a name hierarchy or separator could work, but the `:` character is already taken for specifying subproject option names. Dunno what other character could be used for separation. Maybe `.`?

**Comment by Ericson2314** [comment]
@jpakkane Sure `.` is fine. I proposed `:` *because* it is the existing pass separator: we just need the tiniest something to disambiguate the hierarchy root. For example, there are no nested subprojects today, right? If so any path with 3 or more components is unused today. If we do use `.`, then I'd hope `:` could be deprecated with the new system---I don't care what the separator is, as long the new system doesn't have redundant syntax.

**Comment by jpakkane** [comment]
There should be two different separators because they separate different things. Otherwise there will be namespace clashes. The obvious (stupid) example is that having a subproject called `build` would mean that `build.something` could mean either a "build" option named `something` or the `something` option of a subproject `build`.

**Comment by Ericson2314** [comment]
That's why I was thinking all the hierarchical ones could be 3 deep (e.g. `compiler_options:build:something`) but yeah `.` is fine.

**Comment by Ericson2314** [comment]
With https://github.com/mesonbuild/meson/pull/5414 and surely other "downstream" changes on the horizon, I really hope we can merge this and then do the hierarchical paths, so I'm stuck with fewer rebases / waiting for a release cycle. While I'm sure there are `build_*` user options out there, I doubt anyone is using the specific (`build_`) options this adds.

**Comment by jpakkane** [comment]
I'd say the option name should probably be something like `build.c_args`. And that an option name may have an `.' or a `:` in its name but not both.

**Comment by Ericson2314** [comment]
@jpakkane Are you saying do that in this PR, before a general solution for hierarchical names?

**Comment by Ericson2314** [comment]
Not sure what the conclusion is, but with code cutoff this day I did a rebase to use `build.` instead and fix conflicts.

**Comment by jpakkane** [comment]
This is a lot nicer now. Would be possible to also get `c_args` et al to support `build.c_args` for 0.51.0? And that c args set in the cross file go to `build.c_args` and are exposed properly (when running `meson configure` at least)? It's not mandatory, but it would make for an awesome feature package.

**Comment by Ericson2314** [comment]
> Would be possible to also get c_args et al to support build.c_args for 0.51.0? 

Already done!

> And that c args set in the cross file go to build.c_args and are exposed properly (when running meson configure at least)?

Err... the c args set in the cross file affect the host platform. If we had `[properties]` in the native file it would affect `build.c_args`. I think @dcbaker had a proposal for creating things finer-grained sections than `[properties]` which would go in both files so `[properties]` would be deprecated.

> It's not mandatory, but it would make for an awesome feature package.

:) Glad you think so!

**Comment by jpakkane** [comment]
Err, crap, I got that wrong. Obviously cross c_args to to plain c_args, while CFLAGS et al go to build.c_args.

**Comment by Ericson2314** [comment]
[Speaking of `CFLAGS`, there is also https://github.com/mesonbuild/meson/issues/3969 where`CFLAGS` should affect the host platform, and `CFLAGS_FOR_BUILD` affect the build platform, but one step at a time.]

Do the windows errors make any sense to you, @jpakkane? I am at a bit of a loss what might cause them.

**Comment by jpakkane** [comment]
No more env vars. Things should be defined via cross and native files instead.

> Do the windows errors make any sense to you

They are most likely not related to this PR. Other ones are failing in the same way.

**Comment by Ericson2314** [comment]
I pushed because commit message and docs had `build_` still. Maybe I'll roll the windows CI dice right too!

File: mesonbuild/optinterpreter.py [hunk 15]
```diff
@@ -23,7 +23,7 @@
 
 forbidden_option_names = set(coredata.builtin_options.keys())
 forbidden_prefixes = [lang + '_' for lang in compilers.all_languages] + ['b_', 'backend_']
-reserved_prefixes = ['cross_', 'build_']
+reserved_prefixes = ['cross_', 'build.']
```
Comment (jpakkane) [inline mesonbuild/optinterpreter.py]: `build.` is not actually needed any more. Option names can not have dots in them (I tested this just now and it behaves correctly).

**Comment by jpakkane** [comment]
The one thing that jumped at me is that when printing options it will correctly print `build.c_args` et al but it does _not_ print those core options that have build variants. Maybe there should be a new top level heading `Core options (for build machine)`.

File: mesonbuild/optinterpreter.py (hunk 15 above)
Comment (Ericson2314) [reply mesonbuild/optinterpreter.py]: oh haha! didn't mean to change that line. (I think reserving `build_`) is still a good idea.

**Comment by Ericson2314** [comment]
@jpakkane good point about printing that is definitely an oversight I'll double check all those.

**Comment by Ericson2314** [comment]
Both those things should be fixed!

**Pull request merged**